Communications/PacketBuffer.cs

```diff
@@ -0,0 +1,538 @@
+        /// <param name="Priority"> Should always be zero because there's just one queue. </param>
+        /// 
+        /// <exception cref="ArgumentException"> If priority is not zero since there's just one queue. </exception>
+        public override void Enqueue(Packet Packet, int Priority = 0)
+        {
+            if (Priority != 0)
+                throw new ArgumentException("Leaky bucket controller doesn't have priority.");
+
+            lock (Queue)
+            {
+                Queue.Enqueue(Packet);
+            }
+        }
+
+        /// <summary>
+        /// Get next packet without removing it.
+        /// </summary>
+        /// 
+        /// <remarks>
+        /// It will return null if the buffer is empty instead of throwing an exception.
+        /// </remarks>
+        /// 
+        /// <returns> Next packet if buffer is not empty, or null otherwise. </returns>
+        public override Packet Peek()
+        {
+            lock (Queue)
+            {
+                try
+                {
+                    return Queue.Peek();
+                }
+                catch (InvalidOperationException)
+                {
+                    return null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get next packet and remove it.
+        /// </summary>
+        /// 
+        /// <remarks>
+        /// It will return null if the buffer is empty instead of throwing an exception.
+        /// </remarks>
+        /// 
+        /// <returns> Next packet if buffer is not empty, or null otherwise. </returns>
+        public override Packet Dequeue()
+        {
+            lock (Queue)
+            {
+                try
+                {
+                    return Queue.Dequeue();
+                }
+                catch (InvalidOperationException)
+                {
+                    return null;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the total number of packets in the buffer.
+        /// </summary>
+        public override int Count
+        {
+            get
+            {
+                lock (Queue)
+                {
+                    return Queue.Count;
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// This defines a priority buffer that dequeue the highest priority packet first.
+    /// </summary>
+    public class PriorityBuffer : PacketBuffer
+    {
+        private readonly PacketBuffer[] Buffers; // List of buffers of each priority
+        public readonly int NBuffers; // Number of buffers
+
+        /// <summary>
+        /// Construct priority buffer with a list of sub-buffers.
+        /// </summary>
+        /// 
+        /// <param name="Buffers"> List of buffers for each priority. Low index means higher priority. </param>
+        public PriorityBuffer(PacketBuffer[] Buffers)
+        {
+            this.Buffers = Buffers;
+            NBuffers = Buffers.Length;
```

Comment:
Please make sure to use `this.` for all field access/assignment for consistency.